### PR TITLE
[FLINK-30245][table-runtime] Fix the NPE thrown when filtering decimal(18, x) values after calling DecimalDataUtils.subtract method.

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request fixes the NPE thrown when filtering decimal(18, x) values after calling DecimalDataUtils.subtract method.*

## Brief change log

  - *Check the result precision before calculating the compacted decimal values.*

## Verifying this change

This change is already covered by existing tests, such as *org.apache.flink.table.data.DecimalDataTest.testNotCompact()*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (DecimalData subtract/add only)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/DecimalDataUtils.java [hunk 1]
```diff
@@ -104,7 +105,10 @@ public static DecimalData ceil(DecimalData decimal) {
     }
 
     public static DecimalData add(DecimalData v1, DecimalData v2, int precision, int scale) {
-        if (v1.isCompact() && v2.isCompact() && v1.scale == v2.scale) {
+        if (v1.isCompact()
+                && v2.isCompact()
+                && v1.scale == v2.scale
+                && precision <= MAX_COMPACT_PRECISION) {
```
Comment (luoyuxia) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/DecimalDataUtils.java:111]: Nit: use  `DecimalData.isCompact(precision)`

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/DecimalDataUtils.java [hunk 2]
```diff
@@ -118,7 +122,10 @@ public static DecimalData add(DecimalData v1, DecimalData v2, int precision, int
     }
 
     public static DecimalData subtract(DecimalData v1, DecimalData v2, int precision, int scale) {
-        if (v1.isCompact() && v2.isCompact() && v1.scale == v2.scale) {
+        if (v1.isCompact()
+                && v2.isCompact()
+                && v1.scale == v2.scale
+                && precision <= MAX_COMPACT_PRECISION) {
```
Comment (luoyuxia) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/DecimalDataUtils.java:128]: dito

**Review by luoyuxia — COMMENTED** [review]
@WeiZhong94 Thanks for contribution. Otherwise, LGTM.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/DecimalDataUtils.java (hunk 1 above)
Comment (WeiZhong94) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/DecimalDataUtils.java]: Updated

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/DecimalDataUtils.java (hunk 2 above)
Comment (WeiZhong94) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/DecimalDataUtils.java]: Updated

**Review by luoyuxia — APPROVED** [review]
Thanks for updating. LGTM

**Pull request merged**